### PR TITLE
(DIO-944) First pass at combined web/scraper

### PR DESCRIPTION
**Pull request by jcoconnor**
Setup a gocron task to run the main scraper every minute.

### CODE REVIEW

**Comment by jcoconnor** [comment]
FWIW - this initial bit doesn't go.

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,41 @@
+#
+pipeline {
+    agent any
+
+    triggers { cron("0 * * * *") }
+
+
+    parameters {
+        string(name: 'DB_BRANCH', defaultValue: 'master', description: 'What branch of the pipeline Dashboard should be deployed?')
+        string(name: 'BRANCH', defaultValue: 'master', description: 'What branch of the infra-infra config should be deployed?')
```
Comment (genebean) [inline Jenkinsfile:10]: is this param still needed?

File: frontend/src/App.js [hunk 2]
```diff
@@ -3,15 +3,10 @@ import './App.css';
...
 
-import date from './data_updated.js'
-
 function App() {
   return (
     <div className="App">
       <AppRouter />
-      <footer>
-      Data Last Updated: { date() }
-      </footer>
```
Comment (genebean) [inline frontend/src/App.js:14]: Why remove the update info?

**Review by genebean — COMMENTED** [review]
As a future thing, we should move this Jenkinsfile to running on a k8s worker a la what is done in dio-helm-charts

File: Jenkinsfile (hunk 1 above)
Comment (jcoconnor) [reply Jenkinsfile:10]: Probably not - and there are a few other parameters needed.
Focussing on the actual code first to get it running but need full rework of the Jenksfile

File: frontend/src/App.js (hunk 2 above)
Comment (jcoconnor) [reply frontend/src/App.js:14]: Was removed to get the app running.
Going to replace this with a request to the web app to return the actual updated date.
Up to now it was based on the last Jenkins (deploy run)

File: lib/report/start.go [hunk 3]
```diff
@@ -134,11 +136,15 @@ func ApplyCith(pipelines []jenkins_types.Pipeline, cithFailures []cith.CithFailu
             }
         }
     }
+    utils.Unlock()
```
Comment (Andrew-Hayes) [inline lib/report/start.go:139]: id move this to under the lock (at line 130) and make it a 

`defer utils.Unlock()`

File: lib/report/start.go [hunk 4]
```diff
@@ -134,11 +136,15 @@ func ApplyCith(pipelines []jenkins_types.Pipeline, cithFailures []cith.CithFailu
...
 
     return retVal
 
 }
 
 func WriteToCSV(pipelines []jenkins_types.Pipeline) {
+    // Locks added around file writes to prevent issues with updater.
+    utils.Lock()
     csv_writers.WritePipelines(pipelines)
+    utils.Unlock()
```
Comment (Andrew-Hayes) [inline lib/report/start.go:149]: same again, move up and make defer.

File: lib/web/main.go [hunk 5]
```diff
@@ -114,6 +123,8 @@ func (h *Handlers) GeneratePageData() *Page {
             PlatformVersion:      line[12],
         })
     }
+    trainCSVFile.Close()
+    utils.Unlock()
```
Comment (Andrew-Hayes) [inline lib/web/main.go:127]: again move up and make defer

File: lib/web/main.go [hunk 6]
```diff
@@ -29,6 +30,13 @@ type Page struct {
...
     title := "CI Dashboard"
 
+    fmt.Println("Generating Page Data.")
+
+    // Does file exist - otherwise return.
+
+    // if ()
+
+    utils.Lock()
     csvFile, _ := os.Open("result.csv")
```
Comment (Andrew-Hayes) [inline lib/web/main.go:40]: a good few errors being ignored, i'd log them at least for your own debugging sanity

File: main.go [hunk 7]
```diff
@@ -145,3 +121,25 @@ func main() {
 
     report.WriteToCSV(compiledData)
 }
+
+
+func main() {
+
+    pflag.Bool("hidetreelog", false, "Whether or not to hide the long tree log")
+    pflag.Bool("no-cache", true, "Whether or not to use a cache")
+    pflag.Uint64("scrape-interval", SCRAPE_INTERVAL, "Interval in Seconds to scrape Jenkins/CITH")
```
Comment (Andrew-Hayes) [inline main.go:130]: you could probably set a default for this with viper.

File: lib/report/start.go (hunk 3 above)
Comment (jcoconnor) [reply lib/report/start.go:139]: 👍

Comment (jcoconnor) [reply lib/report/start.go:139]: Marking this and same below as resolved as pushing change up now to reflect that.

File: lib/web/main.go [hunk 8]
```diff
@@ -29,7 +32,20 @@ type Page struct {
...
+    defer utils.Unlock()
+
+    // Does file exist - otherwise return with no page data
+    _, err := os.Stat(constants.Trains_filename)
+    if os.IsNotExist(err) {
+        h.Page = nil
+        return h.Page
+    }
+
+    csvFile, _ := os.Open(constants.Results_filename)
```
Comment (scottyw) [inline lib/web/main.go:48]: Best to check the `err` here and below.

File: lib/web/main.go [hunk 9]
```diff
@@ -68,8 +84,9 @@ func (h *Handlers) GeneratePageData() *Page {
             Transients: transients,
         })
     }
+    csvFile.Close()
```
Comment (scottyw) [inline lib/web/main.go:87]: Ideally you'd `defer` this immediately after opening the file without error. Same below.

File: lib/report/start.go [hunk 10]
```diff
@@ -137,7 +137,6 @@ func ApplyCith(pipelines []jenkins_types.Pipeline, cithFailures []cith.CithFailu
             }
         }
     }
-    utils.Unlock()
```
Comment (scottyw) [inline lib/report/start.go]: Aye, I think you fixed it but @Andrew-Hayes's pattern of `defer-ing` the unlock as soon as you lock helps avoid this.

**Review by scottyw — COMMENTED** [review]
Regarding the threading errors til the files are ready - I guess you can safely just try to open them, check for an error, backoff, try agin for some number of iterations. That's seems alright to me.

File: frontend/src/App.js (hunk 2 above)
Comment (jcoconnor) [reply frontend/src/App.js:14]: Latest commit replaces this with a new field in the products data returned from the Go App.

File: lib/web/main.go (hunk 9 above)
Comment (jcoconnor) [reply lib/web/main.go:88]: Done - incoming.

File: lib/web/main.go (hunk 8 above)
Comment (jcoconnor) [reply lib/web/main.go:49]: K - done

File: main.go (hunk 7 above)
Comment (jcoconnor) [reply main.go]: Do you mean move aaway from pflag altogether?

File: lib/web/main.go (hunk 6 above)
Comment (jcoconnor) [reply lib/web/main.go]: Think I'm checking all of these now.

File: main.go (hunk 7 above)
Comment (jcoconnor) [reply main.go]: Need to get this out, so going for a simple approach of putting all the config parametes in the .TOML file.
Don't want to change the existing structures, but do need to drive to a better solution in a later revision.

**Comment by jcoconnor** [comment]
I'm going to declare this one ready for review/merge as I want to do the Jenkinsfile Update in a seperate PR.

**Review by genebean — APPROVED** [review]
Based on slack chats and looking at the non-go bits, this seems good to me.

**Pull request merged**